File: trade-monitor/metrics.go
package main

import (
    "github.com/prometheus/client_golang/prometheus"
    "github.com/prometheus/client_golang/prometheus/promauto"
)

var (
    // Метрики для мониторинга сервисов
    serviceHealth = promauto.NewGaugeVec(
        prometheus.GaugeOpts{
            Name: "service_health",
            Help: "Health status of services (1 = healthy, 0 = unhealthy)",
        },
        []string{"service"},
    )

    serviceUptime = promauto.NewGaugeVec(
        prometheus.GaugeOpts{
            Name: "service_uptime_seconds",
            Help: "Service uptime in seconds",
        },
        []string{"service"},
    )

    // Метрики для торговых операций
    tradeProcessingTime = promauto.NewHistogramVec(
        prometheus.HistogramOpts{
            Name: "trade_processing_seconds",
            Help: "Time spent processing trade",
        },
        []string{"symbol", "side"},
    )

    tradeProcessingErrors = promauto.NewCounterVec(
        prometheus.CounterOpts{
            Name: "trade_processing_errors_total",
            Help: "Total number of trade processing errors",
        },
        []string{"error_type"},
    )

    tradeVolume = promauto.NewCounterVec(
        prometheus.CounterOpts{
            Name: "trade_volume_total",
            Help: "Total trading volume",
        },
        []string{"symbol", "side"},
    )

    activeTrades = promauto.NewGaugeVec(
        prometheus.GaugeOpts{
            Name: "active_trades",
            Help: "Number of active trades",
        },
        []string{"symbol"},
    )

    // Метрики для Kafka
    kafkaConsumerLag = promauto.NewGaugeVec(
        prometheus.GaugeOpts{
            Name: "kafka_consumer_lag",
            Help: "Kafka consumer lag",
        },
        []string{"topic", "partition"},
    )

    kafkaMessagesProcessed = promauto.NewCounterVec(
        prometheus.CounterOpts{
            Name: "kafka_messages_processed_total",
            Help: "Total number of Kafka messages processed",
        },
        []string{"topic"},
    )

    // Метрики для алертов
    alertCount = promauto.NewCounterVec(
        prometheus.CounterOpts{
            Name: "alert_count_total",
            Help: "Total number of alerts generated",
        },
        []string{"severity", "type"},
    )

    // Метрики для производительности
    messageProcessingLatency = promauto.NewHistogramVec(
        prometheus.HistogramOpts{
            Name:    "message_processing_latency_seconds",
            Help:    "Message processing latency in seconds",
            Buckets: prometheus.DefBuckets,
        },
        []string{"type"},
    )

    // Метрики для базы данных
    dbQueryTime = promauto.NewHistogramVec(
        prometheus.HistogramOpts{
            Name: "db_query_seconds",
            Help: "Time spent executing database queries",
        },
        []string{"query_type"},
    )

    dbConnectionErrors = promauto.NewCounter(
        prometheus.CounterOpts{
            Name: "db_connection_errors_total",
            Help: "Total number of database connection errors",
        },
    )

    // Метрики для API
    apiRequestTime = promauto.NewHistogramVec(
        prometheus.HistogramOpts{
            Name: "api_request_seconds",
            Help: "Time spent processing API requests",
        },
        []string{"endpoint", "method"},
    )

    apiErrors = promauto.NewCounterVec(
        prometheus.CounterOpts{
            Name: "api_errors_total",
            Help: "Total number of API errors",
        },
        []string{"endpoint", "status_code"},
    )
)

// Функции-помощники для обновления метрик
func updateServiceHealth(service string, healthy bool) {
    if healthy {
        serviceHealth.WithLabelValues(service).Set(1)
    } else {
        serviceHealth.WithLabelValues(service).Set(0)
    }
}

func updateServiceUptime(service string, uptime float64) {
    serviceUptime.WithLabelValues(service).Set(uptime)
}

func updateKafkaLag(group, topic string, lag float64) {
    kafkaConsumerLag.WithLabelValues(group, topic).Set(lag)
}

func incrementMessagesProcessed(topic string) {
    kafkaMessagesProcessed.WithLabelValues(topic).Inc()
}

func incrementAlertCount(severity, alertType string) {
    alertCount.WithLabelValues(severity, alertType).Inc()
} 

File: trade-executor/metrics.py
from prometheus_client import Counter, Gauge, Histogram
import time

# Метрики для торговых операций
trade_executions = Counter(
    'trade_executions_total',
    'Total number of trade executions',
    ['symbol', 'side', 'status']
)

trade_latency = Histogram(
    'trade_execution_latency_seconds',
    'Trade execution latency in seconds',
    ['symbol']
)

active_trades = Gauge(
    'active_trades',
    'Number of active trades',
    ['symbol', 'side']
)

trade_pnl = Gauge(
    'trade_pnl',
    'Current PnL for trades',
    ['symbol', 'side']
)

# Метрики для баланса
account_balance = Gauge(
    'account_balance',
    'Current account balance',
    ['currency']
)

margin_ratio = Gauge(
    'margin_ratio',
    'Current margin ratio',
    ['currency']
)

# Метрики для Kafka
kafka_messages_sent = Counter(
    'kafka_messages_sent_total',
    'Total number of messages sent to Kafka',
    ['topic']
)

kafka_message_latency = Histogram(
    'kafka_message_latency_seconds',
    'Kafka message processing latency in seconds',
    ['topic']
)

# Метрики для ошибок
trade_errors = Counter(
    'trade_errors_total',
    'Total number of trade execution errors',
    ['error_type']
)

# Метрики для Prometheus
TRADE_PROCESSING_TIME = Histogram(
    'trade_processing_seconds',
    'Time spent processing trade',
    ['symbol', 'side']
)

TRADE_PROCESSING_ERRORS = Counter(
    'trade_processing_errors_total',
    'Total number of trade processing errors',
    ['error_type']
)

TRADE_VOLUME = Counter(
    'trade_volume_total',
    'Total trading volume',
    ['symbol', 'side']
)

# Метрики для Kafka
KAFKA_MESSAGES_PROCESSED = Counter(
    'kafka_messages_processed_total',
    'Total number of Kafka messages processed',
    ['topic']
)

KAFKA_CONSUMER_LAG = Gauge(
    'kafka_consumer_lag',
    'Kafka consumer lag',
    ['topic', 'partition']
)

# Метрики для базы данных
DB_QUERY_TIME = Histogram(
    'db_query_seconds',
    'Time spent executing database queries',
    ['query_type']
)

DB_CONNECTION_ERRORS = Counter(
    'db_connection_errors_total',
    'Total number of database connection errors'
)

# Метрики для API
API_REQUEST_TIME = Histogram(
    'api_request_seconds',
    'Time spent processing API requests',
    ['endpoint', 'method']
)

API_ERRORS = Counter(
    'api_errors_total',
    'Total number of API errors',
    ['endpoint', 'status_code']
)

# Декоратор для измерения латентности
def measure_latency(metric):
    def decorator(func):
        def wrapper(*args, **kwargs):
            start_time = time.time()
            try:
                result = func(*args, **kwargs)
                metric.observe(time.time() - start_time)
                return result
            except Exception as e:
                trade_errors.labels(error_type=type(e).__name__).inc()
                raise
        return wrapper
    return decorator 